src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -0,0 +1,160 @@
+package de.tum.in.www1.artemis.service;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.*;
+import de.tum.in.www1.artemis.domain.lecture.ExerciseUnit;
+import de.tum.in.www1.artemis.domain.lecture.LectureUnit;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+import de.tum.in.www1.artemis.web.rest.dto.LearningGoalProgress;
+
+@Service
+public class LearningGoalService {
+
+    private final ParticipationService participationService;
+
+    public LearningGoalService(ParticipationService participationService) {
+        this.participationService = participationService;
+    }
+
+    /**
+     * Calculates the progress of the given user in the given exercise units
+     *
+     * Note: In the case of two exercise units referencing the same exercise, only the first exercise unit will be used.
+     *
+     * Note: Please note that we take always the last submission into account here. Even submissions after the due date. This means for example that a student can improve his/her
+     * progress by re-trying a quiz as often as you like. It is therefore normal, that the points here might differ from the points officially achieved in an exercise.
+     *
+     * @param exerciseUnits exercise units to check
+     * @param user user to check for
+     * @return progress of the user in the exercise units
+     */
+    public Set<LearningGoalProgress.LectureUnitProgress> calculateExerciseUnitsProgress(Set<ExerciseUnit> exerciseUnits, User user) {
+        // for each exercise unit, the exercise will be mapped to a freshly created lecture unit progress.
+        Map<Exercise, LearningGoalProgress.LectureUnitProgress> exerciseToLectureUnitProgress = exerciseUnits.stream().filter(exerciseUnit -> exerciseUnit.getExercise() != null)
+                .filter(exerciseUnit -> exerciseUnit.getExercise().isAssessmentDueDateOver()).collect(Collectors.toMap(ExerciseUnit::getExercise, exerciseUnit -> {
```

Comment:
here you could combine both filters into one:  

.filter(exerciseUnit -> exerciseUnit.getExercise() != null && exerciseUnit.getExercise().isAssessmentDueDateOver()).collect(...

Reply:
c70aa0f01060fd36a9f540c34c4b5f0e009a4410